hypothesis-python/src/hypothesis/internal/scrutineer.py

```diff
@@ -38,11 +39,45 @@ def trace(self, frame, event, arg):
         if event == "call":
             return self.trace
         elif event == "line":
-            fname = frame.f_code.co_filename
-            if should_trace_file(fname):
-                current_location = (fname, frame.f_lineno)
-                self.branches.add((self._previous_location, current_location))
-                self._previous_location = current_location
+            self.trace_line(frame.f_code, frame.f_lineno)
+
+    def trace_line(self, code: types.CodeType, line_number: int):
+        fname = code.co_filename
+        if should_trace_file(fname):
+            current_location = (fname, line_number)
+            self.branches.add((self._previous_location, current_location))
+            self._previous_location = current_location
+
+
+# where possible, we'll use 3.12's new sys.monitoring module for low-overhead
+# coverage instrumentation. Otherwise, we'll default to sys.settrace.
+# This can be simplified once we drop 3.11.
+# tool_id = 2 is designated for coverage.
+MONITORING_TOOL_ID = 2
```

Comment:
I've used the tool id that [PEP 669](https://peps.python.org/pep-0669/) reserves for coverage, as scrutineer is ostensibly collecting coverage information. If/when coverage.py switches to using monitoring as well - which may happen in the near future https://github.com/nedbat/coveragepy/issues/1665#issuecomment-1778042967 - our tool ids will conflict[^1]. This might be a nonissue, as the current code only runs when there is no other trace function (read: coverage.py isn't running) anyway, so we could add a check for tool id 2 being free and call it a day. 

Alternatively, we could choose an unreserved tool ID, and gain the ability to run scrutineer at the same time as coverage.py. I'm not actually sure whether this is desired, however.

[^1]: unless they choose not to use ID 2, but I can't imagine why they wouldn't.

Reply:
I think we'll want to use one of the unclaimed tool IDs here, since I'd perfer to be compatible with coverage measurement if that's feasible (and I think it now is).